File: Konfigurator-aspnetcore-Backend/Konfigurator-aspnetcore-Backend/Services/Game/GameService.cs
using Konfigurator_aspnetcore_Backend.Data;
using Konfigurator_aspnetcore_Backend.Models;
using Konfigurator_aspnetcore_Backend.Models.DTOs;
using Microsoft.EntityFrameworkCore;

namespace Konfigurator_aspnetcore_Backend.Services.Game
{
    public class GameService : IGameService
    {
        private readonly LabyrinthKonfiguratorContext _context;
        public GameService(LabyrinthKonfiguratorContext context)
        {
            _context = context;
        }
        public async Task<ServiceResponseDto<GetGameMapResponseDto>> GetGameMap(int gameMapId)
        {
            // Daten aus Datenbank auslesen
            var gameMap = await _context.gameMaps
                .Include(gm => gm.barrierDesign)
                .Include(gm => gm.characterDesign)
                .Include(gm => gm.targetDesign)
                .Include(gm => gm.tileDesign)
                .Include(gm => gm.obstaclesInMaps)
                .Include(gm => gm.highscores)
                .FirstOrDefaultAsync(gm => gm.gameMapId == gameMapId);

            // wurde keine GameMap zur gegebenen gameMapId gefunden
            if(gameMap == null)
            {
                return new ServiceResponseDto<GetGameMapResponseDto> { Success = false, Message = "gesuchte GameMap nicht vorhanden" };
            }

            /*
             * Werte mappen
             */
            var obstaclesInMap = new List<VectorDto>();

            foreach(var element in gameMap.obstaclesInMaps)
            {
                obstaclesInMap.Add(new VectorDto(element.x, element.y));
            }

            var getGameMapDto = new GetGameMapDto(
                gameMap.gameMapId,
                gameMap.name,
                gameMap.username,
                new MapSizeDto(gameMap.sizeX, gameMap.sizeY),
                gameMap.tile,
                gameMap.barrier,
                gameMap.character,
                gameMap.target,
                obstaclesInMap,
                new VectorDto(gameMap.startPosX, gameMap.startPosY),
                new VectorDto(gameMap.endPosX, gameMap.endPosY));

            var response = new GetGameMapResponseDto(getGameMapDto);

            return new ServiceResponseDto<GetGameMapResponseDto> { Data = response, Success = true };
        }

        public async Task<ServiceResponseDto<GetHighscoresDto[]>> GetHighscores(int gameMapId)
        {
            var highscores = await _context.highscores.ToArrayAsync();

            List<GetHighscoresDto> response = new List<GetHighscoresDto>();

            foreach (Highscore score in highscores)
            {
                if(score.gameMapId == gameMapId)
                {
                    response.Add(new GetHighscoresDto(score.date, score.gameMapId, score.username, score.score));
                }
            }

            if(response.Count == 0)
            {
                return new ServiceResponseDto<GetHighscoresDto[]> { Success = false };
            }
            var toret = response.ToArray();
            return new ServiceResponseDto<GetHighscoresDto[]> { Data = toret, Success = true };
        }

        public async Task<ServiceResponse> AddHighscore(int gameMapId, string username, int score)
        {
            Highscore newScore = new Highscore();

            newScore.date = DateTime.Now;
            newScore.username = username;
            newScore.gameMapId = gameMapId;
            newScore.score = score;

            await _context.highscores.AddAsync(newScore);
            await _context.SaveChangesAsync();

            return new ServiceResponse { Success = true, Message = "Score: " + score + " von: " + username + " zu map: " + gameMapId + " in Datenbank gespeichert" };
        }



        public async Task<ServiceResponseDto<GetAllGameMapIdDto>> GetAllGameMapIds()
        {
            // Daten aus Datenbank auslesen
            var gameMaps = await _context.gameMaps.ToArrayAsync();

            // wurde keine GameMaps gefunden
            if (gameMaps.Length == 0)
            {
                return new ServiceResponseDto<GetAllGameMapIdDto> { Success = false, Message = "keine GameMaps vorhanden" };
            }

            /*
             * Werte mappen
             */

            List<int> ids = new List<int>();

            foreach (GameMap gm in gameMaps)
            {
                if (gm != null)
                {
                    ids.Add(gm.gameMapId);
                }
            }

            var response = new GetAllGameMapIdDto(ids.ToArray());

            return new ServiceResponseDto<GetAllGameMapIdDto> { Data = response, Success = true };
        }

        public async Task<ServiceResponseDto<byte[]>> GetGraphic(string graphic)
        {
            var response = new ServiceResponseDto<byte[]> { Success = true };

            // alle vier Tabellen mit Grafiken nach der gesuchten Grafik durchsuchen
            var barrier = await _context.barrierDesigns.FindAsync(graphic);
            if (barrier != null)
            {
                response.Data = barrier.img;
                return response;
            }

            var character = await _context.characterDesigns.FindAsync(graphic);
            if(character != null)
            {
                response.Data = character.img;
                return response;
            }

            var target = await _context.targetDesigns.FindAsync(graphic);
            if(target != null)
            {
                response.Data = target.img;
                return response;
            }

            var tile = await _context.tileDesigns.FindAsync(graphic);
            if(tile != null)
            {
                response.Data = tile.img;
                return response;
            }

            // nicht gefunden
            response.Success = false;
            response.Message = "Grafik nicht gefunden";
            return response;
        }
    }
}

File: Konfigurator-nestjs-Backend/konfigurator-nestjs-backend/src/game/game.service.ts
import { Injectable } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm'
import { Repository } from 'typeorm'
import { ServiceResponseDto } from 'src/dto/ServiceResponse.dto';
import { GetGameMapDto } from 'src/dto/get-GameMap.dto';
import { GetGameMapResponseDto } from 'src/dto/get-GameMapResponse.dto';
import { GetHighscoresDto } from '../dto/get-Highscores.dto'
import { GameMaps } from 'src/entities/GameMapsEntity';
import { Highscores } from 'src/entities/HighscoresEntity';
import { VectorDto } from 'src/dto/Vector.dto';
import { MapSizeDto } from 'src/dto/MapSize.dto';
import { GetAllGameMapIdDto } from 'src/dto/get-AllGameMapId.dto';
import { BarrierDesigns } from 'src/entities/BarrierDesignsEntity';
import { CharacterDesigns } from 'src/entities/CharacterDesignsEntity';
import { TargetDesigns } from 'src/entities/TargetDesignsEntity';
import { TileDesigns } from 'src/entities/TileDesignsEntity';
import { ObstaclesInMaps } from 'src/entities/ObstaclesInMapsEntity';

@Injectable()
export class GameService {
  constructor(
    @InjectRepository(GameMaps) private gameMapsRepository: Repository<GameMaps>,
    @InjectRepository(Highscores) private highscoresRepository: Repository<Highscores>,
    @InjectRepository(BarrierDesigns) private barrierDesignsRepository: Repository<BarrierDesigns>,
    @InjectRepository(CharacterDesigns) private characterDesignsRepositroy: Repository<CharacterDesigns>,
    @InjectRepository(TargetDesigns) private targetDesignsRepositroy: Repository<TargetDesigns>,
    @InjectRepository(TileDesigns) private tileDesignsRepositroy: Repository<TileDesigns>,
    @InjectRepository(ObstaclesInMaps) private ObstaclesInMapsRepository: Repository<ObstaclesInMaps>
  ) {}

  async GetHighscores(gameMapId: number) {

    let highscores = await this.highscoresRepository.find();
    let highscoresDtos: GetHighscoresDto[] = [];

    // Fill highscore array
    highscores.forEach(score =>{
      if(score.gameMapId == gameMapId)
      {
        let newScore: GetHighscoresDto = new GetHighscoresDto(score.date,score.gameMapId,score.username,score.score); 
        highscoresDtos.push(newScore);
      }
    })

    if(highscoresDtos.length == 0)
      return new ServiceResponseDto<GetHighscoresDto[]>(null,false,"Keine Highscores gefunden");
    else
      return new ServiceResponseDto<GetHighscoresDto[]>(highscoresDtos,true,"Highscores verfügbar");
  }

  async GetGameMap(gameMapId: number) {

    let gameMap = await this.gameMapsRepository.find()
    let obstaclesInMaps = await this.ObstaclesInMapsRepository.find();

    let obstaclesInMapdto: VectorDto[] = [];
    let getGameMapDto: GetGameMapDto;

    if(gameMap == null)
      return new ServiceResponseDto<GetGameMapResponseDto[]>(null,false,"Keine GameMap gefunden");
    
    /*
     * Werte mappen
     */
    obstaclesInMaps.forEach(element => {
      if(element.gameMapId == gameMapId)
        obstaclesInMapdto.push(new VectorDto(element.x,element.y))
    })

    gameMap.forEach(element => {
      if(element.gameMapId == gameMapId)
      {
        let newsize = new MapSizeDto(element.sizeX,element.sizeY);
        let startpos = new VectorDto(element.startPosX, element.startPosY);
        let endpos = new VectorDto(element.endPosX, element.endPosY);

        let newgamemap: GetGameMapDto = new GetGameMapDto(
          element.gameMapId,
          element.name,
          element.username,
          newsize,
          element.tile,
          element.barrier,
          element.character,
          element.target,
          obstaclesInMapdto,
          startpos,
          endpos);
          getGameMapDto = newgamemap;
      }
    })

    let response = new GetGameMapResponseDto(getGameMapDto);

    return new ServiceResponseDto<GetGameMapResponseDto>(response,true,"Game Map verfügbar")
  }

  async GetAllGameMapsIds() {

    let gameMaps = await this.gameMapsRepository.find();
    let id: number[] = []
    // let GameMapIdDto: GetAllGameMapIdDto[] = [];

    gameMaps.forEach(gm =>{
      if (gm != null)
      {
        id.push(gm.gameMapId)
      }
    })

    let GameMapIdDto = new GetAllGameMapIdDto(id);
    
    if (gameMaps.length == 0)
      return new ServiceResponseDto<GetAllGameMapIdDto[]>(null,false,"Keine Ids verfügbar");
    else
      return new ServiceResponseDto<GetAllGameMapIdDto>(GameMapIdDto,true,"Gamemap Ids verfügbar");
  }

  async GetGraphic(graphic: string) {

    let barrierDesigns = await this.barrierDesignsRepository.find()
    let characterDesigns = await this.characterDesignsRepositroy.find();
    let targetDesigns = await this.targetDesignsRepositroy.find();
    let tileDesigns = await this.tileDesignsRepositroy.find();

    let data: any = null;

    barrierDesigns.forEach(element => {
      if(element.barrier == graphic) 
      {
        data = element.img.toString('base64');
      }
    });

    characterDesigns.forEach(element => {
      if(element.character == graphic) 
      {
        data = element.img.toString('base64');
      }
    });

    targetDesigns.forEach(element => {
      if(element.target == graphic) 
      {
        data = element.img.toString('base64');
      }
    });

    tileDesigns.forEach(element => {
      if(element.tile == graphic) 
      {
        data = element.img.toString('base64');
      }
    });

    if(data != null)
      return new ServiceResponseDto<Buffer[]>(data,true,"Grafik verfügbar")
    else
      return new ServiceResponseDto<Buffer[]>(data,false,"Grafik nicht verfügbar")

  }

  async addHighscore(gameMapId: number, username: string, score: number) {

    let highscores = new Highscores;

    highscores.date = new Date().toString();
    highscores.gameMapId = gameMapId;
    highscores.score = score;
    highscores.username = username;

    await this.highscoresRepository.save(highscores);

    return new ServiceResponseDto<string>("",true,"Hinzufügen erfolgt");


  }
}
